File: OrangeJuice.Server.Core/Services/XmlItemSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;

namespace OrangeJuice.Server.Services
{
	public sealed class XmlItemSelector : IItemSelector
	{
		private readonly IValidator<XElement> _itemValidator;

		public XmlItemSelector(IValidator<XElement> itemValidator)
		{
			_itemValidator = itemValidator;
		}

		public XElement[] SelectItems(string xml)
		{
			XDocument doc = XDocument.Parse(xml);
			XNamespace ns = doc.Root.Name.Namespace;

			XElement items = doc.Root.Element(ns + "Items");
			if (!_itemValidator.IsValid(items))
				throw new ArgumentException(GetErrorMessage(doc, ns));

			return items.Elements(ns + "Item").ToArray();
		}

		private static string GetErrorMessage(XDocument doc, XNamespace ns)
		{
			var nm = new XmlNamespaceManager(new NameTable());
			nm.AddNamespace("x", ns.NamespaceName);

			return doc.XPathSelectElement("//x:Error", nm).Value;
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/XmlItemSelector.fs
namespace OrangeJuice.Server.FSharp.Services

open System
open System.Collections.Generic
open System.Xml
open System.Xml.Linq
open System.Xml.XPath

open OrangeJuice.Server
open OrangeJuice.Server.Services

type XmlItemSelector(itemValidator : IValidator<XElement>) =
    interface IItemSelector with
        member this.SelectItems(xml : string) : IEnumerable<XElement> =
            let doc = XDocument.Parse(xml)
            let ns = doc.Root.Name.Namespace

            let items = doc.Root.Element(ns + "Items")
            if itemValidator.IsValid(items) then items.Elements(ns + "Item")
            else raise <| new ArgumentException(XmlItemSelector.GetErrorMessage(doc, ns))

    static member GetErrorMessage(doc : XDocument, ns : XNamespace ) : string =
        let nm = new XmlNamespaceManager(new NameTable())
        nm.AddNamespace("x", ns.NamespaceName) |> ignore
        doc.XPathSelectElement("//x:Error", nm).Value
